### PR TITLE
Tests to verify 'startup_tsa_tsb.service' during linecard bring up - PR#18496

### PR DESCRIPTION
**Pull request by sanjair-git**
### Description of PR

Summary:
Fixes # (issue)
- This PR has tests to verify the changes introduced as part of #[18496](https://github.com/sonic-net/sonic-buildimage/pull/18496)
- Basically, this PR has 5 test cases to verify start and stop of the new service '_startup_tsa_tsb.service_' added by #[18496](https://github.com/sonic-net/sonic-buildimage/pull/18496) including other checks.

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [x] 202205
- [x] 202305
- [x] 202311

#### What is the motivation for this PR?
- To verify code changes introduced as part of the PR #[18496](https://github.com/sonic-net/sonic-buildimage/pull/18496)

#### How did you do it?

- Added 5 test cases as mentioned below to verify the changes. 
- Attaching test cases' description document here for further reference and details.
[BGP-TSA-TSB-Service-TestCases.txt](https://github.com/sonic-net/sonic-mgmt/files/14808766/BGP-TSA-TSB-Service-TestCases.txt)

1. _test_tsa_tsb_service_with_dut_cold_reboot_
2. _test_tsa_tsb_service_with_dut_abnormal_reboot_
3. _test_tsa_tsb_service_with_supervisor_cold_reboot_
4. _test_tsa_tsb_service_with_supervisor_abnormal_reboot_
5. _test_tsa_tsb_service_with_user_init_tsa_

#### How did you verify/test it?

- Ran all the above mentioned test cases on a T2 chassis and made sure tests passed with expected behavior.

### Documentation

image

### CODE REVIEW

File: tests/bgp/test_startup_tsa_tsb_service.py [hunk 1]
```diff
@@ -0,0 +1,725 @@
...
+                duthosts, linecard, dut_nbrhosts[linecard], traffic_shift_community),
+                "Failed to verify routes on nbr in TSA")
+
+        # Once all line cards are in maintenance state, proceed further
+        for linecard in duthosts.frontend_nodes:
+            # Verify startup_tsa_tsb service stopped after expected time
+            pytest_assert(wait_until(tsa_tsb_timer[linecard], 20, 0, get_tsa_tsb_service_status, linecard, 'exited'),
+                          "startup_tsa_tsb service is not stopped even after configured timer expiry")
+
+            pytest_assert(wait_until(90, 5, 0, check_tsc_command_error, linecard),
```
Comment (tjchadaga) [inline tests/bgp/test_startup_tsa_tsb_service.py:586]: why is this required? In what cases does TSC error out after TSA-TSB timer expiry?

Comment (sanjair-git) [reply tests/bgp/test_startup_tsa_tsb_service.py]: Not required actually. Added it for debugging purposes initially. Removed it now.

File: tests/bgp/test_startup_tsa_tsb_service.py [hunk 2]
```diff
@@ -0,0 +1,725 @@
...
+        pytest_assert(wait_until(tsa_tsb_timer, 20, 0, get_tsa_tsb_service_status, duthost, 'exited'),
+                      "startup_tsa_tsb service is not stopped even after configured timer expiry")
+
+        # Ensure dut comes back to normal state after timer expiry
+        if not get_tsa_tsb_service_status(duthost, 'running'):
+            # Verify TSB is configured on the dut after startup_tsa_tsb service is stopped
+            pytest_assert(TS_NORMAL == get_traffic_shift_state(duthost),
+                          "DUT is not in normal state after startup_tsa_tsb service is stopped")
+        # Make sure DUT continues to be in good state after TSB
+        assert wait_until(300, 20, 2, duthost.critical_services_fully_started), \
```
[the remark this answers is not in the document]
Comment (sanjair-git) [reply tests/bgp/test_startup_tsa_tsb_service.py]: There is no need to wait more here after tsa-tsb timer expires. Removed the code.

**Comment by sanjair-git** [comment]
> @sanjair-git - could we add testcases to validate the behavior when user issues TSA/TSB while the service is running? This could be from linecard or supervisor
> 
> Also, a case to validate whether the tsa-tsb timer is sufficient for the system to be stable (ports up, critical services, bgp state etc)

@tjchadaga, For the new test cases, I will raise a separate PR. Other comments, I have looked into it and resolved. Reran the whole test suite again with the changes. Results below,

image

**Review by tjchadaga — APPROVED** [review]

**Review by tjchadaga — APPROVED** [review]

**Review by abdosi — APPROVED** [review]
